"make push" should have consistent behavior

**Describe the feature request**

`make push` does entirely different things depending on whether your remote repo (HUB variable) is set to GCR or not. If it's non-GCR, you get a docker push, otherwise it uploads artifacts to GCS. This is implemented as a condition in the makefile that [re-writes the push rule.](https://github.com/istio/istio/blob/1e062cf7108d254d929be8d22561f5856ddab9b8/Makefile#L627-L628)

This creates an awkward surprising behavior for devs that aren't expecting it. 

I've already updated the initially confusing doc (https://github.com/istio/istio/wiki/Using-the-Code-Base) to say `make docker.push` instead, which gives the expected behavior in that particular context... but this seems like something that should be fixed more broadly. Is there a reason for the current behavior? It seems very strange to have `make push` do entirely separate and orthogonal things based on the value of the HUB variable. Given how many people are using GCR to hold their test images, it's a recipe for confusion. 

A simple proposal: what if we got rid of the condition that rewrites `make push` and instead created a new rule to handle the "push to GCS" case? Maybe `make release` or `make artifacts`?  

Would this approach break anything? I'm not sure what workflows rely on the current (strange) behavior. [A cursory search](https://github.com/istio/istio/search?q=%22make+push%22&unscoped_q=%22make+push%22) makes it look some test & deploy automation is involved, although it certainly seems like the automation expects the "docker push" behavior instead of the "gcs push" behavior. (Perhaps the conditional rewrite in the makefile is platform dependent, and it doesn't affect the automated workflow?)


**Describe alternatives you've considered**

We could also just clean up documentation so that the behavior is well-explained and use `docker.push` anywhere that's actually what we want to happen. 

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[X] Test and Release
[ ] User Experience
[X] Developer Infrastructure


**Additional context**

Repro steps on my dev host:
```console
crwilson@crwilson:~/go/src/istio.io/istio$ HUB=bogus make push
time (mkdir -p /usr/local/google/home/crwilson/go/out/linux_amd64/release/docker_build/docker.pilot && cp -r pilot/docker/Dockerfile.pilot /usr/local/google/home/crwilson/go/out/linux_amd64/release/pilot-discovery tests/testdata/certs/cacert.pem /usr/local/google/home/crwilson/go/out/linux_amd64/release/docker_build/docker.pilot && cd /usr/local/google/home/crwilson/go/out/linux_amd64/release/docker_build/docker.pilot &&  docker build  --build-arg BASE_DISTRIBUTION=default -t bogus/pilot:6b1f464f21e34d15156d76fd7ad488e366b84980 -f Dockerfile.pilot . ); 
Sending build context to Docker daemon  71.41MB
Step 1/7 : ARG BASE_DISTRIBUTION=default
Step 2/7 : FROM istionightly/base_debug as default
 ---> d484448b52e3
Step 3/7 : FROM gcr.io/distroless/static as distroless
 ---> a574914f27fd
Step 4/7 : FROM ${BASE_DISTRIBUTION}
 ---> d484448b52e3
Step 5/7 : ADD pilot-discovery /usr/local/bin/
 ---> Using cache
 ---> 8c73c68e3194
Step 6/7 : ADD cacert.pem /cacert.pem
 ---> Using cache
 ---> f2b95b93f654
Step 7/7 : ENTRYPOINT ["/usr/local/bin/pilot-discovery"]
 ---> Using cache
 ---> 24d94d583aa3
Successfully built 24d94d583aa3
Successfully tagged bogus/pilot:6b1f464f21e34d15156d76fd7ad488e366b84980

real	0m2.688s
user	0m1.496s
sys	0m0.771s
time (set -e && for distro in default; do tag=6b1f464f21e34d15156d76fd7ad488e366b84980-${distro}; docker push bogus/pilot:${tag%-default}; done)
The push refers to repository [docker.io/bogus/pilot]
6003260952b6: Preparing 
9e9706afabf7: Preparing 
02748e0fc4e2: Preparing 
4c54072a5034: Preparing 
49652298c779: Preparing 
e15278fcccca: Waiting 
739482a9723d: Waiting 
denied: requested access to the resource is denied

real	0m1.218s
user	0m0.026s
sys	0m0.038s
tools/istio-docker.mk:259: recipe for target 'push.docker.pilot' failed
make: *** [push.docker.pilot] Error 1
crwilson@crwilson:~/go/src/istio.io/istio$ HUB=gcr.io/bogus make push
mkdir -p /usr/local/google/home/crwilson/go/out/logs
STATIC=0 GOOS=linux  LDFLAGS='-extldflags -static -s -w' bin/gobuild.sh /usr/local/google/home/crwilson/go/out/linux_amd64/release/istioctl-linux ./istioctl/cmd/istioctl

real	0m0.849s
user	0m1.432s
sys	0m0.511s
STATIC=0 GOOS=darwin LDFLAGS='-extldflags -static -s -w' bin/gobuild.sh /usr/local/google/home/crwilson/go/out/linux_amd64/release/istioctl-osx ./istioctl/cmd/istioctl

real	0m0.828s
user	0m1.364s
sys	0m0.492s
STATIC=0 GOOS=windows LDFLAGS='-extldflags -static -s -w' bin/gobuild.sh /usr/local/google/home/crwilson/go/out/linux_amd64/release/istioctl-win.exe ./istioctl/cmd/istioctl

real	0m0.795s
user	0m1.337s
sys	0m0.439s
gsutil -m cp -r "/usr/local/google/home/crwilson/go/out/linux_amd64/release"/istioctl-* "gs://istio-artifacts/pilot/6b1f464f21e34d15156d76fd7ad488e366b84980/artifacts/istioctl"
Copying file:///usr/local/google/home/crwilson/go/out/linux_amd64/release/istioctl-linux [Content-Type=application/octet-stream]...
Copying file:///usr/local/google/home/crwilson/go/out/linux_amd64/release/istioctl-osx [Content-Type=application/octet-stream]...
Copying file:///usr/local/google/home/crwilson/go/out/linux_amd64/release/istioctl-win.exe [Content-Type=application/x-msdos-program]...
ResumableUploadAbortException: 403 crwilson@google.com does not have storage.objects.create access to istio-artifacts.
ResumableUploadAbortException: 403 crwilson@google.com does not have storage.objects.create access to istio-artifacts.
ResumableUploadAbortException: 403 crwilson@google.com does not have storage.objects.create access to istio-artifacts.
CommandException: 3 files/objects could not be transferred.
Makefile:633: recipe for target 'gcs.push.istioctl-all' failed
make: *** [gcs.push.istioctl-all] Error 1

```

--- a/Makefile
+++ b/Makefile
@@ -652,15 +652,14 @@ clean.go: ; $(info $(H) cleaning...)
 #-----------------------------------------------------------------------------
 # Target: docker
 #-----------------------------------------------------------------------------
-.PHONY: artifacts gcs.push.istioctl-all artifacts installgen
+.PHONY: push gcs.push gcs.push.istioctl-all gcs.push.deb artifacts installgen
 
 # for now docker is limited to Linux compiles - why ?
 include tools/istio-docker.mk
 
-# if first part of URL (i.e., hostname) is gcr.io then upload istioctl and deb
-$(if $(findstring gcr.io,$(firstword $(subst /, ,$(HUB)))),$(eval push: gcs.push.istioctl-all gcs.push.deb),)
-
 push: docker.push installgen
+
+gcs.push: push gcs.push.istioctl-all gcs.push.deb
 
 gcs.push.istioctl-all: istioctl-all
 	gsutil -m cp -r "${ISTIO_OUT}"/istioctl-* "gs://${GS_BUCKET}/pilot/${TAG}/artifacts/istioctl"

--- a/prow/istio-presubmit.sh
+++ b/prow/istio-presubmit.sh
@@ -42,5 +42,5 @@ if [[ -n $(git diff) ]]; then
 fi
 
 # Upload images - needed by the subsequent tests
-time ISTIO_DOCKER_HUB="gcr.io/istio-testing" make push HUB="gcr.io/istio-testing" TAG="${GIT_SHA}"
+time ISTIO_DOCKER_HUB="gcr.io/istio-testing" make gcs.push HUB="gcr.io/istio-testing" TAG="${GIT_SHA}"
 

